wit/readwrite.wit

```diff
@@ -1,26 +1,56 @@
 /// A keyvalue interface that provides simple read and write operations.
+/// 
+/// A read/write operation is an operation that acts on a single key-value pair.
+/// 
+/// The value in the key-value pair is defined as a `u8` byte array and the intention
+/// is that it is the common denominator for all data types defined by different
+/// key-value stores to handle data, ensuring compatibility between different
+/// key-value stores. Note: the clients will be expecting serialization/deserialization overhead
+/// to be handled by the key-value store. The value could be a serialized object from
+/// JSON, HTML or vendor-specific data types like AWS S3 objects.
+/// 
+/// Data consistency in a key value store refers to the gaurantee that once a 
+/// write operation completes, all subsequent read operations will return the
+/// value that was written.
+/// 
+/// The level of consistency in readwrite interfaces is **eventual consistency**,
+/// which means that if a write operation completes successfully, all subsequent
+/// read operations will eventually return the value that was written. In other words,
+/// if we pause the updates to the system, the system eventually will return 
+/// the last updated value for read.
 interface readwrite {
```

Comment:
I like this suggestion. What do you think @Mossaka?

Reply:
I am a bit concenred about the over-specification risk that could lead to an API that's too rigid. Also terms like `eventual` may be interpreted differently in different contexts. For example, what does it mean to use `wasi:keyvalue/eventual` in a single-node scenario? In addition, it assumes the users of these interfaces must have an understanding of the consistency models in distributed systems, which may not be the case. With that said, I don't have a strong opinion against the proposal, I myself not sure about the balance between clarity vs. complexity here. 

There is also an interesting question about WASI interface naming. Do we want the interface name to reflect more on the behavior of the system (e.g. `eventual`), or more on the types of the operations (e.g. `readwrite`)?

Reply:
To make things a bit more consistent, I would like to keep `readwrite` on the naming of the interface. Maybe somerthing like 

- `wasi:keyvalue/eventual-readwrite`?

Reply:
I think there is no avoiding stating the required consistency as part of the semantic contract.  It adds complexity, but non-strong consistency is just inherently complicated -- the naming is the least of the problems.  As I was discussing earlier, if you try to ignore consistency, you'll just get broken developer expectations leading to bugs that only show up late.

> For example, what does it mean to use wasi:keyvalue/eventual in a single-node scenario?

You don't need to be distributed to have weak (eventual) consistency: this can happen even when you have two threads racing on the same memory in the same process, so even a single-process server implemented in terms of relaxed atomics can exhibit weak consistency.

The issue I'm perceiving with "readwrite" is that most key-value interfaces are doing some combination of reading and writing (and deleting...).  `wasi:keyvalue/atomic` is also doing reading and writing (and probably moreso in the future); it's just doing so *with sequentially-consistent (atomic) consistency*.  Calling it `wasi:keyvalue/eventual` was thus meant to be consistent with `wasi:keyvalue/atomic`.  But maybe we could pick a common suffix, like `-ops` so that it'd be `wasi:keyvalue/atomic-ops`/`wasi:keyvalue/eventual-ops`; I dunno if that makes the interface names more readable or just longer?

Reply:
> there is no avoiding stating the required consistency as part of the semantic contract

I agree, based on our previous discussion.

> You don't need to be distributed to have weak (eventual) consistency

I have mixed feeling. On the one side, I get what you are saying - When multiple threads are involved in a single-process, writes by one thread might not immediately be visible to other threads due to factors like caching, compiler optimizations, or the reordering of instructions. On the other side, to my knowledge, this scenario is more often categorize under data synchronization or concurrency control mechanisms, where memory barrier or lock is used to ensure data synchronization. Perhaps this's not the traditional use of the term _eventual consistency_. A traditional meaning of eventual consistency is defined under async replicated systems where data is copied to >1 nodes asynchronously. Given this, I would think that people might be confused about what `wasi:keyvalue/eventual` means for using the interface in their single-node environment. 

I think we are trying to find a term that describes a "weaker guarantee" than **atomic** under both distributed or single-node systems. I don't know what's the best one, but I think `eventual` is a good one. 

>  Calling it wasi:keyvalue/eventual was thus meant to be consistent with wasi:keyvalue/atomic

I checked how [Dapr's State Management](https://docs.dapr.io/developing-applications/building-blocks/state-management/state-management-overview/) name things and I noticed that they have a clear concept seperation not only for data consistency, but also for concurrency control. Dapr supports 2 concurrency control mechanisms: "last-write-wins" and "optimistic concurrency control (OCC) with etags". It also supports 2 consistency control: "eventual" and "strong". Given that we are discussing adding consistency level to the interface names, what do you think about concurrency control as another dimension that may be part of the semantic contract?

Reply:
To move this forward, my suggestion is to leave the interface name as is. When we start designing "strong" consistent APIs, we can look back and investigate exactly how the function signatures are different in these two consistency levels. 

It may well be that we only need to add an extra parameter to each of the CRUD operations in `readwrite` interface to indicate what consistency level the user expects (or to the `bucket` resource to configure a scoped consistency level) and this will avoid an extra interfce specifically designed for that consistency level. 

It may be not, and some signature changes force us to add an extra interface for that respective consistency level. Then it makes perfect sense to change the `readwrite` interface name to `eventual`.

Reply:
(sorry for the slow reply! just got back)

> Given that we are discussing adding consistency level to the interface names, what do you think about concurrency control as another dimension that may be part of the semantic contract?

"OCC with etags" is an interesting option.  IIUC, this approach adds an extra etag result and parameter to most operations so it seems like you'd naturally want to define a new `interface` to capture this new functionality.  I think this is a good example of how, in general, different consistency options will want different sets of operations and function signatures and thus, in this early phase when we're "setting the mold", it's a good idea to preemptively partition operations into interfaces based on consistency.  And again, having this choice be part of the interface name and thus statically part of the component signature helps developers and platforms know, before runtime, what a component needs to run (or if the component can run at all, since not all consistency options are available on all stores).

> To move this forward, my suggestion is to leave the interface name as is.

It sounded like you were pretty on-board with `wasi:keyvalue/eventual`; could we just do that?  While it's a superficial naming thing that we could break in Preview 3, I think in this case the naming will help convey our future design direction (which, as our discussions show, isn't obvious from first principles, and thus valuable to help folks toward).

Reply:
Done!